File: process_sync_semaphores.c
#include <fcntl.h>
#include <semaphore.h>
#include <stdio.h>
#include <sys/wait.h>
#include <unistd.h>
int main() {
sem_t *semaphore;
pid_t pid;
char *sem_name = "/my_semaphore";
semaphore = sem_open(sem_name, O_CREAT | O_EXCL, 0666, 0);
if (semaphore == SEM_FAILED) {
perror("Creating semaphore failed");
return 1;
}
pid = fork();
if (pid < 0) {
perror("Fork failed");
return 1;
} else if (pid == 0) {
printf("Child waiting for semaphore\n");
sem_wait(semaphore);
printf("Child acquired semaphore\n");
// Critical section starts
sleep(5);
// Critical section ends
sem_post(semaphore);
} else {
sleep(1);
printf("Parent releasing semaphore\n");
sem_post(semaphore);
wait(NULL);
}
sem_close(semaphore);
sem_unlink(sem_name);
return 0;
}

File: process_sync_semaphores.py
import posix_ipc
import os
import time
sem_name = "/my_semaphore"
try:
semaphore = posix_ipc.Semaphore(sem_name, posix_ipc.O_CREAT | posix_ipc.O_EXCL, 0o666, 0)
except posix_ipc.ExistentialError:
raise Exception("Creating semaphore failed")
pid = os.fork()
if pid < 0:
raise Exception("Fork failed")
elif pid == 0:
print("Child waiting for semaphore")
semaphore.acquire()
print("Child acquired semaphore")
# Critical section starts
time.sleep(5)
# Critical section ends
semaphore.release()
semaphore.close() # Close the semaphore in the child process
else:
time.sleep(1)
print("Parent releasing semaphore")
semaphore.release()
os.waitpid(pid, 0)
semaphore.close() # Close the semaphore in the parent process
# Unlink the semaphore after closing it
posix_ipc.unlink_semaphore(sem_name)
